Add glossary of terms to documentation (#3210)
pallets/click: docs/glossary.rst

Glossary
========

This glossary defines key terms used throughout Click's documentation.

.. glossary::
   :sorted:

   Argument
       A positional parameter passed to a command without a name prefix.
       Arguments are identified by their position in the command line.
       For example, in ``mycli command file.txt``, ``file.txt`` is an argument.

   Callback
       A function that is invoked when a parameter is processed. Callbacks can be
       attached to options or arguments for custom validation or processing.
       See :doc:`/parameters` for more information.

   Command
       A function decorated with ``@click.command()`` or ``@click.group()``
       that can be invoked from the command line. Commands can accept options
       and arguments.

>    Command Line Application
>        A Python program built with Click that provides a command line interface.
>        Command line applications use commands, options, and arguments to interact
>        with users through the terminal.
> 
>    Command Line Interface (CLI)
>        A text-based interface for interacting with a program through commands,
>        options, and arguments. Click helps you build CLIs in Python.

Reviewer: Correct me if I'm wrong here. But I feel like "CLI" is common nomenclature for the application, regardless of what the true acronym stands for.

```suggestion
   Command-line Application
       Application meant to be interacted with using a command-line interface.
       Often referred to as a CLI as well, even though it's not the true acronym. 

   Command-line Interface (CLI)
       A text-based interface for interacting with a program through commands,
       options, and arguments. While CLI is its acronym, a CLI often refers to
       an application written for the command-line interface.
```

>    Command Line Utility
>        A command line application designed to perform a specific task or set of
>        tasks. Command line utilities are often used for system administration,
>        file manipulation, or development workflows.
> 

Reviewer: ```suggestion
```

Personally I'd remove this to avoid confusion.

   Context
       An object that carries state between commands in a group hierarchy.
       The context is created for each command invocation and can be accessed
       via the ``@click.pass_context`` decorator.

   Flag

>        A boolean option that doesn't take a value (e.g., ``--verbose`` or ``-v``).

Reviewer: I'm not too familiar with how glossaries work in Sphinx. But will "option" here link to the glossary for that word down below automatically or does that need to be done explicitly here and elsewhere in the code? I hope it's the latter.

       Flags are either present (True) or absent (False).

   Group
       A command that contains other commands (subcommands). Created with
       ``@click.group()``. Groups allow organizing commands hierarchically.

> 
>    Multi-command
>        A command that groups multiple subcommands together. Groups and
>        multi-commands are similar concepts in Click.

Reviewer: ```suggestion

   Multi-command
       Alias for Group, used to have a slightly different implementation.
       Planned for removal in 9.0. 
```
